lib/compress/zstd_compress.c

```diff
@@ -3103,6 +3104,12 @@ static size_t ZSTD_buildBlockEntropyStats_sequences(seqStore_t* seqStorePtr,
     size_t entropyWorkspaceSize = wkspSize - (MaxSeq + 1) * sizeof(*countWorkspace);
     ZSTD_symbolEncodingTypeStats_t stats;
 
+    if (nbSeq == 0) {
+        /* Do not build sequences statistics if nbSeq is 0, just return a default */
+        fseMetadata->llType = fseMetadata->mlType = fseMetadata->ofType = set_compressed;
```

Comment:
What are the consequences of settings these `fseMedata` to `set_compressed` ?
We know that there is no sequence, so it's not going to be used in this block.
It also will not be written into the compressed frame.

But what about the _next_ block ?
Will it consider the previous block as featuring "compressed sequences", and therefore entropy headers ?
Asking as it _might_ impact the algorithm which tries to decide between "repeating previous stats" or "issuing new stats".

Reply:
So basically `regressiontest` only passes with the default set to `set_compressed` or `set_rle`.
There are three places where the `encodingType` is read:
https://github.com/facebook/zstd/blob/e7e4b74b2b99016234e8a44b04764ab6305fe83c/lib/compress/zstd_compress_superblock.c#L193-L194
https://github.com/facebook/zstd/blob/e7e4b74b2b99016234e8a44b04764ab6305fe83c/lib/compress/zstd_compress_superblock.c#L368-L369
https://github.com/facebook/zstd/blob/e7e4b74b2b99016234e8a44b04764ab6305fe83c/lib/compress/zstd_compress_superblock.c#L405-L407

The first two cases do not ever get read, due to `nbSeq == 0`. The last case still does, so basically whatever we decide to set `fseMetadata->llType` to, it's basically to affect how this function returns. It seems like even with `nbSeq == 0`, we'd want `ZSTD_needSequenceEntropyTables()` to evaluate to true right? At least the fuzzer tests seem to indicate so.

Reply:
If my understanding is correct, `ZSTD_needSequenceEntropyTables()` is invoked in only a single place :  
https://github.com/facebook/zstd/blob/e7e4b74b2b99016234e8a44b04764ab6305fe83c/lib/compress/zstd_compress_superblock.c#L521

This place seems to check if there is a discrepancy between needing entropy tables, and writing entropy tables.

This feels weird to me.
If `nbSeq==0`, then there should be no need to write any sequence entropy header.
Hence, we should have `writeSeqEntropy == 0`.

But it's set to `1` unconditionally at the beginning of the function.
The only place where it could be updated to `0` is here, and that's for an unrelated reason : 
https://github.com/facebook/zstd/blob/e7e4b74b2b99016234e8a44b04764ab6305fe83c/lib/compress/zstd_compress_superblock.c#L512

This code is very complex, and difficult to follow.
It looks to me that there should be no need to write Sequences Entropy is there is no sequence.
So my first idea would be to change that initialization to `0` in this case.
Unfortunately, the "number of Sequences" is not provided in a straightforward manner,
but is accessible through `(unsigned)(send-sstart)` (discovered from the traces) .

Since the code is complex, it could be that I misinterpret the objective or hidden usages of this variable.
Given that it's then used in multiple places (`ZSTD_estimateSubBlockSize()` and `ZSTD_compressSubBlock()`) 
it feels like a dangerous change.
Worth testing, and maybe worth following what happens to this variable in these sub-functions.

Reply:
Yeah the code is confusing.

Actually, in both `ZSTD_estimateSubBlockSize()` and `ZSTD_compressSubBlock()`, for the `nbSeq == 0` case, we don't use `writeSeqEntropy`. In both cases, we return early if `nbSeq == 0`, leaving that variable unread.

So really, I think it's just this decision of whether or not to emit uncompressed that we end up affecting. 

`writeSeqEntropy == 0` if `seqEntropyWritten = 1`, so we can think of `writeSeqEntropy == 1` as saying "entropy has not been written". So basically the check of 
` if (writeSeqEntropy && ZSTD_needSequenceEntropyTables(&entropyMetadata->fseMetadata))` is saying that if we did not write seqEntropyTables, but did need them, we should emit uncompressed. So my setting of `llType` to `set_compressed/rle` is not correct, since we're saying that we always need them.

But at the same time, setting `writeSeqEntropy = (send-sstart) == 0 ? 0 : 1;` doesn't seem to produce the correct results either, as it fails the tests.

Reply:
It's a pretty bad situation.
It seems we don't understand well enough what's happening in this unit,
which becomes a maintenance liability.
And since it has never worked correctly, it is not used anywhere.
I'm wondering if we would be better off without it.

Reply:
One thing I've figured out is that the test failures are a result of the added early return in `ZSTD_buildBlockEntropyStats_sequences()` for `nbSeq==0` causing us to not run `ZSTD_buildSequencesStatistics()`, which causes us to miss out on some side effects which are apparently necessary - it doesn't actually matter what the `llType, etc.` are when `nbSeq == 0`. I'll modify the PR accordingly to see if I can get that working.

Reply:
I've updated this with a fix that should make sense, though yeah, I'm not sure whether the cost of maintaining this code is worth its general complexity, especially considering that I don't think anyone really uses it.

The added branch should be hardwired for the typical compression path - I measured no speed difference at level 1.